students/alex_skrn/lesson05/mailroom.py

```diff
@@ -0,0 +1,244 @@
+
+
+def get_total_given(name):
+    """Return total amount of donations for the given donor."""
+    return sum(get_donations(name))
+
+
+def sort_donors_by_total():
+    """Return a list of donor names sorted by total donations, max to min."""
+    donors_L = list(donors.items())
+    donors_sorted = sorted(donors_L, key=lambda x: sum(x[1]), reverse=True)
+    sorted_donor_names = [item[0] for item in donors_sorted]
+
+    return sorted_donor_names
+
+
+def print_donor_names():
+    """Print existing donor names on screen."""
+    print()
+    donors_L = list(donors.keys())
+    for name in donors_L[:-1]:
+        print(name, end=', ')
+    print(donors_L[-1])
+    print()
+
+
+def get_email(name, amount):
+    """Return a str containing a thank-you email."""
+    email_text = ("""\nDear {},\n
+                  \nI would like to thank you for your donation of ${}.\n
+                  \nWe appreciate your support.\n
+                  \nSincerely,\n
+                  \nThe Organization\n
+                  """)
+    return email_text.format(name, amount)
+
+
+def print_email(name, amount):
+    """Print a thank-you email on screen."""
+    print(get_email(name, amount))
+
+
+def create_report():
+    """Create a report."""
+    title_line_form = "{:<26}{:^3}{:>13}{:^3}{:>13}{:^3}{:>13}"
+    title_line_text = ('Donor Name', '|', 'Total Given', '|',
+                       'Num Gifts', '|', 'Average Gift'
+                       )
+    print()
+    print(title_line_form.format(*title_line_text))
+    print('- ' * 38)
+    form_line = "{:<26}{:>3}{:>13}{:>3}{:>13}{:>3}{:>13}"
+    for name in sort_donors_by_total():
+        total = get_total_given(name)
+        num_gifts = len(get_donations(name))
+        mean = round((total / num_gifts), 2)
+        print(form_line.format(str(name), '$', str(total), ' ',
+                               str(num_gifts), '$', str(mean)
+                               )
+              )
+    print()
+
+
+# PRINT ON SREEN A THANK YOU LETTER TO SOMEONE WHO JUST MADE A DONATION
+def input_donation(name):
+    """Obtain the donation amount from the user."""
+    prompt_amount = "Enter the donation amount or 0 to abort > "
+    while True:
+        try:
+            donation_amount = float(input(prompt_amount))
+        except ValueError:
+            print("Input must be a number")
+            continue
+        else:
+            if donation_amount == 0.0:
+                return False
+            else:
+                add_donation(name, donation_amount)
+                return True
+    return False
+
+
+def existing_donor_interaction():
+    """Ask for old donor name, donation amount, print a thank-you email."""
+    prompt_name = "Type full name of the old donor or 0 to abort > "
+    old_donor_name = input(prompt_name)
+    if old_donor_name == "0":
+        return
+    while old_donor_name not in donors:
+        old_donor_name = input(prompt_name)
+        if old_donor_name == "0":
+            return
```

Comment:
I think you could re-write the above more simply:
```
old_donor_name = input("Type full name of the old donor or 0 to abort > ")
while old_donor_name not in donors:
     if old_donor_name == "0": return
```

Reply:
Thank you for your comments. I tried to implement this solution but it doesn't seem to work the same way as my function does now. With the proposed solution, if the user enters a wrong name, then the function seems to enter into an infinite while loop without giving the user an oppurtunity to re-enter a name. On the other hand, if I add one more input statement inside the while loop so that the user gets such an opportunity to re-enter a name, than it makes it impossible for the user to enter 0 on the first (i.e. outer) prompt and quit immediately.  Instead, the flow goes into the while loop and the user is prompted once again to type 0 or a name.

Reply:
```
    old_donor_name = ''
    while old_donor_name not in donors:
        old_donor_name = input("Type full name of the old donor or 0 to abort > ")
        if old_donor_name == "0": return 
```

Reply:
Very cool, thanks!